### ISSUE TITLE
Memory leakage when calling multiple instances in a loop

### ISSUE THREAD

**Issue by akshaysravindran** [jacobgil/pytorch-grad-cam#132]
Hi,

Thank you for releasing this package, it is very useful for the community. However, I noticed an issue associated with memory leakage when I call multiple instances of the CAM object. When I reinitiate the CAM object in a loop, the memory is cumulatively being built up even after I delete the object at the end of the loop. I require recreating the CAM object for the application I have in a loop and cannot reuse the same object defined once.  To replicate the issues, I am adding a small code snippet. Do you have any thoughts on why this might be happening and how I could avoid it?

Thanks,

`import psutil

## Define the model and train it
input = torch.randn(16, 1, 62, 250, requires_grad=True)
labels = np.zeros((16,1))

for i in range(100):
    cam     = GradCAM(model=model, target_layer=model.conv5,use_cuda=True)
    grad    =  cam(input_tensor=input, target_category=labels)    
    
    print(psutil.cpu_percent())
    print(psutil.virtual_memory())  # physical memory usage
    print('memory % used:', psutil.virtual_memory()[2])    
    del cam,grad`

**Comment by jacobgil** [comment]
Hi,
My initial guess is that the ActivationsAndGradients isn't deleted, and the hooks aren't un-registered.
I am looking into it and will update.

**Comment by jacobgil** [comment]
Should be solved!
https://github.com/jacobgil/pytorch-grad-cam/pull/133
Once the tests pass I will merge.

**Comment by jacobgil** [comment]
The internal ActivationsAndGradients was not releasing hooks.
I ended up creating a .release() function for it,
and call it from the cam object __del__ function.
This means you should be able to just use it as is, in a loop, without leakage.

But since we can't really rely on when exactly __del__ is called, 
the CAM objects now also support the with statement.

You can now explicitly do
```python
for i in range(1000):
    with cam_method(model=model, target_layers=target_layers) as cam:
        cam_output = cam(input_tensor=input, target_category=None)
```

**Comment by jacobgil** [comment]
It's merged into the master branch.
Does it work for you now?

**Comment by jacobgil** [comment]
Closing this issue, it should be solved now, please re-open if you see the problem continues.

**Comment by akshaysravindran** [comment]
Yes, it is working now! Thanks for the prompt action, this is super helpful!

**Comment by AntixK** [comment]
Hi @jacobgil ,

I think the problem still persists. I have attached the output from running `GradCam` model on the features of DenseNet161 on a custom dataset.

```
| Feature Extraction | Test |:   0%|                    | 0/169 [00:00<?, ?it/s]% Memory used: 36.5
| Feature Extraction | Test |:   1%|▍                   | 1/169 [00:20<56:12, 20.07s/it]% Memory used: 37.5
| Feature Extraction | Test |:   1%|▊                   | 2/169 [00:36<49:13, 17.68s/it]% Memory used: 38.3
| Feature Extraction | Test |:   2%|█▎                  | 3/169 [00:49<43:48, 15.84s/it]% Memory used: 38.7
| Feature Extraction | Test |:   2%|█▋                  | 4/169 [01:04<42:33, 15.48s/it]% Memory used: 39.6
| Feature Extraction | Test |:   3%|██▏                 | 5/169 [01:22<44:28, 16.27s/it]% Memory used: 40.3
| Feature Extraction | Test |:   4%|██▌                 | 6/169 [01:37<43:11, 15.90s/it]% Memory used: 40.9
| Feature Extraction | Test |:   4%|███                 | 7/169 [01:52<42:29, 15.74s/it]% Memory used: 41.7
| Feature Extraction | Test |:   5%|███▍                | 8/169 [02:07<40:55, 15.25s/it]% Memory used: 42.6
| Feature Extraction | Test |:   5%|███▉                | 9/169 [02:20<38:48, 14.55s/it]% Memory used: 43.5
| Feature Extraction | Test |:   6%|████▎               | 10/169 [02:32<36:39, 13.84s/it]% Memory used: 44.0
| Feature Extraction | Test |:   7%|████▋               | 11/169 [02:45<35:54, 13.64s/it]% Memory used: 44.6
| Feature Extraction | Test |:   7%|█████               | 12/169 [03:03<38:46, 14.82s/it]% Memory used: 45.5
| Feature Extraction | Test |:   8%|█████▌              | 13/169 [03:17<38:29, 14.80s/it]% Memory used: 46.4
| Feature Extraction | Test |:   8%|█████▉              | 14/169 [03:34<39:37, 15.34s/it]% Memory used: 47.0
| Feature Extraction | Test |:   9%|██████▍             | 15/169 [03:49<39:10, 15.26s/it]% Memory used: 47.8
| Feature Extraction | Test |:   9%|██████▊             | 16/169 [04:05<39:14, 15.39s/it]% Memory used: 48.7
| Feature Extraction | Test |:  10%|███████▏            | 17/169 [04:19<38:00, 15.00s/it]% Memory used: 49.2
| Feature Extraction | Test |:  11%|███████▋            | 18/169 [04:32<36:16, 14.41s/it]% Memory used: 49.8
| Feature Extraction | Test |:  11%|████████            | 19/169 [04:49<37:49, 15.13s/it]
```

Clear, the memory used is increasing gradually. For, reference, the following is the code I use to extract the gradcam matrix

```
with GradCAM(model=model, 
                   target_layers=target_layers['layers'], 
                   use_cuda=True if device != 'cpu' else False, ) as cam:
    
    grayscale_cam = cam(input_tensor=input_tensor,
                            target_category=None,
                            eigen_smooth=True,
                            aug_smooth=True)
```

**Comment by AntixK** [comment]
I have figured out the problem. Just commenting here in case it might be of help to others. 

I don't think this is an issue of `pytorch_grad_cam`. This is an issue with the way I wrapped the PyTorch model. A similar issue has been reported in the `PyTorch` repo - https://github.com/pytorch/pytorch/issues/46386#issue-722155792

Python's gc fails to clean up if the `nnModule` object is wrapped with another class. So, when using hooks, it is better not to wrap the model to avoid memory leaks.
